src/drivers/amd/rembrandt/acp_sw_audio_dma.c

```diff
@@ -0,0 +1,640 @@
+//
+//Author:       Basavaraj Hiregoudar <basavaraj.hiregoudar@amd.com>
+//              Maruthi Machani <maruthi.machani@amd.com>
+
+#include <rtos/atomic.h>
+#include <sof/audio/component.h>
+#include <rtos/bit.h>
+#include <sof/drivers/acp_dai_dma.h>
+#include <rtos/interrupt.h>
+#include <rtos/timer.h>
+#include <rtos/alloc.h>
+#include <rtos/clk.h>
+#include <sof/lib/cpu.h>
+#include <sof/lib/dma.h>
+#include <sof/lib/io.h>
+#include <sof/lib/pm_runtime.h>
+#include <sof/lib/notifier.h>
+#include <sof/platform.h>
+#include <sof/schedule/schedule.h>
+#include <rtos/spinlock.h>
+#include <sof/math/numbers.h>
+#include <sof/trace/trace.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <platform/fw_scratch_mem.h>
+#include <platform/chip_registers.h>
+
+/*b414df09-9e31-4c59-8657-7afc8deba70c*/
+DECLARE_SOF_UUID("acp-sw-audio", acp_sw_audio_uuid, 0xb414df09, 0x9e31, 0x4c59,
+		 0x86, 0x57, 0x7a, 0xfc, 0x8d, 0xeb, 0xa7, 0x0c);
+DECLARE_TR_CTX(acp_sw_audio_tr, SOF_UUID(acp_sw_audio_uuid), LOG_LEVEL_INFO);
+
+#define SW0_AUDIO_FIFO_SIZE			128
+#define SW0_AUDIO_TX_FIFO_ADDR		0
+#define SW0_AUDIO_RX_FIFO_ADDR		(SW0_AUDIO_TX_FIFO_ADDR + SW0_AUDIO_FIFO_SIZE)
+
+#define SW0_BT_FIFO_SIZE		128
+#define SW0_BT_TX_FIFO_ADDR		(SW0_AUDIO_RX_FIFO_ADDR + SW0_AUDIO_FIFO_SIZE)
+#define SW0_BT_RX_FIFO_ADDR		(SW0_BT_TX_FIFO_ADDR + SW0_BT_FIFO_SIZE)
+
+#define SW0_HS_FIFO_SIZE		128
+#define SW0_HS_TX_FIFO_ADDR		(SW0_BT_RX_FIFO_ADDR + SW0_BT_FIFO_SIZE)
+#define SW0_HS_RX_FIFO_ADDR		(SW0_HS_TX_FIFO_ADDR + SW0_HS_FIFO_SIZE)
+
+#define SW1_FIFO_SIZE			128
+#define SW1_TX_FIFO_ADDR		(SW0_HS_RX_FIFO_ADDR + SW1_FIFO_SIZE)
+#define SW1_RX_FIFO_ADDR		(SW1_TX_FIFO_ADDR + SW1_FIFO_SIZE)
+
+static uint32_t sw_audio_buff_size_playback;
+static uint32_t sw_audio_buff_size_capture;
+
+struct sw_dev_register {
+	uint32_t	sw_dev_en;
+	uint32_t	sw_dev_en_status;
+	uint32_t	sw_dev_fifo_addr;
+	uint32_t	fifo_addr;
+	uint32_t	sw_dev_fifo_size;
+	uint32_t	fifo_size;
+	uint32_t	sw_dev_ringbuff_addr;
+	uint32_t	sw_dev_ringbuff_size;
+	uint32_t	sw_dev_dma_size;
+	uint32_t	sw_dev_dma_watermark;
+	uint32_t	sw_dev_dma_intr_status;
+	uint32_t	sw_dev_dma_intr_cntl;
+	uint32_t	statusindex;
+};
+
+struct sw_dev_register sw_dev[8];
+
+/* allocate next free DMA channel */
+static struct dma_chan_data *acp_dai_sw_audio_dma_channel_get(struct dma *dma,
+							      unsigned int req_chan)
+{
+	k_spinlock_key_t key;
+	struct dma_chan_data *channel;
+
+	key = k_spin_lock(&dma->lock);
+	if (req_chan >= dma->plat_data.channels) {
+		k_spin_unlock(&dma->lock, key);
+		tr_err(&acp_sw_audio_tr, "Channel %d not in range", req_chan);
+		return NULL;
+	}
+	channel = &dma->chan[req_chan];
+	if (channel->status != COMP_STATE_INIT) {
+		k_spin_unlock(&dma->lock, key);
+		tr_err(&acp_sw_audio_tr, "channel already in use %d", req_chan);
+		return NULL;
+	}
+	atomic_add(&dma->num_channels_busy, 1);
```

Comment:
num_channels_busy is atomic.

Reply:
@maruthi-machani I see that, and that's exactly my question - since it's always protected by a spinlock, it doesn't have to be an atomic variable, a simple integer would be enough

Reply:
@lyakh yes, int would be good enough, need to check as modifying this would affect other platforms of NXP and Mediatek.